### ISSUE TITLE
Feature request: inherit Kodi's audio device

### ISSUE THREAD

**Issue by tuzz** [wackerl91/script.luna#94]
- Luna Version: 0.6.3
- nVidia GeForce Experience Version: 3.1.2.31
- Moonlight Embedded Version: 2.2.2
- OSMC Version: 2016.10-1

**Feature request:**

Hi, thanks for building a great plugin. I was wondering if it would be possible to inherit the audio device in use by Kodi. I'm using a USB audio device and have it set up correctly in Kodi. I haven't been able to get it to work (yet) in Luna and it would really help if the two were in sync.

Thanks

**Comment by wackerl91** [comment]
A possible solution for this depends on two things:
- does Luna's audio selection (or starting moonlight manually with the '-audio' option and proper parameter) work for you? I've said this numerous times on different occasions, the audio selection in Luna has been a feature request and I don't have the means to actually test it, but I never received any feedback on this.
- if there is a way to get the current audio device via the Kodi API (which I simply don't know off the top of my head)

**Comment by wackerl91** [comment]
@tuzz 
Played around with Kodi's JSONRPC API for a bit ... it looks like the only (relevant) value I can get about the currently selected audio device is the name shown in the UI, e.g.:
```
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "value": "PI:Analogue"
  }
}
```
I could hardcode some known special cases, like Pi: Analogue translating to bcm2835 ALSA and Pi: HDMI translating to (probably, correct me if I'm wrong) bcm2835 IEC958/HDMI.

All other devices will be problematic, since their representation will most likely be different from what information I can obtain by parsing /proc/asound/cards (I can't use aplay since ALSA utils won't be installed on every system and I can't make this a hard requirement obviously).

For example I know for a fact that a Logitech G35 Headset will be present as both:
* ALSA: Logitech G35 Headset, Analog
* ALSA: Logitech G35 Headset, S/PDIF

aplay will list a lot of devices for this headset, all being to the same card / device (1:0 in my case). Parsing /proc/asound/cards (and some other files in those folders) will yield the same information as well in this case. So I could just look for 'Logitech G35 Headset' (ignoring the Analog / S/PDIF part) and most likely select the proper device.
This isn't necessarily true for all cases though, especially for USB sound cards having multiple subdevices, so it'd be more like guessing which device you selected for Kodi. Of course, partial matches would be possible (find all devices named similarly, look for a substring (e.g. 'HDMI') and return the best match, aka 'best guess'), but that's too much work for too little benefit. Additionally, I can already see an implementation like the above going wrong all the time.

If you have any idea for a proper and reliable implementation - without relying on additional packages being installed - I'm happy to think about it again (or you could do it yourself and open a PR).
Until then I'd rather work on making the existing audio device selection better / more user friendly.

Sorry for the 'bad news', I still appreciate the idea and I'm looking forward to your feedback.

**Comment by tuzz** [comment]
Hi @wackerl91, thanks for your detailed response.

I don't have a huge amount of knowledge to offer in this regard, but I can share my thoughts.

Firstly, it's a shame you can't retrieve that information from Kodi's APIs. I would have thought there are other plugins that produce audio, but I guess that's different because they're not passing a specific device identifier to a third-party program.

I was wondering how exhaustive the list of devices in `/proc/asound/cards` is? Does `aplay -l` and `aplay -L` include additional devices that aren't listed there? If I'm able to set my audio device in Kodi (which might not have ALSA utils), surely it must be in `/proc/asound/cards` ?

You mentioned that a G35 headset would be listed twice. I found this too for my audio device. In fact, both the Analog and S/PDIF outputs work. I think the Analog device has slightly less latency. I don't know if that's typical for all audio devices or not.

In summary, I think this is the right thing to do:

>I'd rather work on making the existing audio device selection better / more user friendly.

Having a dropdown that gives you a full list of devices would be great.

Sorry I can't help more. Thanks a lot.

**Comment by wackerl91** [comment]
Believe me, I'd love to get some more detailed information from Kodi :-/
I can paste the output of aplay -l, aplay -L and /proc/asound/cards while the G35 is attached, but I'd rather do so tomorrow / on Monday as it's kinda late right now.
I guess my issue with this will be obvious once you have a look at those outputs.

As a side note: while trying to find a solution for your FR I already corrected some issues with the device listing. I was able to select my G35 and it was used by moonlight properly. What I still cannot do is switch between analog and hdmi audio on the Broadcom chip (might just be my setup though).
So Luna's audio device selection seems to be working in some cases ... I'm sure we can get your end working as well, there's still some stuff left we could try.

(P.S.: I had a look at your BSc thesis the other night, really fun stuff! And keep up all the other good work you're doing!)

**Comment by wackerl91** [comment]
@tuzz
Warning, huge wall of text ;)

aplay -l
```
**** List of PLAYBACK Hardware Devices ****
card 0: ALSA [bcm2835 ALSA], device 0: bcm2835 ALSA [bcm2835 ALSA]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 0: ALSA [bcm2835 ALSA], device 1: bcm2835 ALSA [bcm2835 IEC958/HDMI]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: Headset [Logitech G35 Headset], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

aplay -L
```
null
    Discard all samples (playback) or generate zero samples (capture)
default:CARD=ALSA
    bcm2835 ALSA, bcm2835 ALSA
    Default Audio Device
sysdefault:CARD=ALSA
    bcm2835 ALSA, bcm2835 ALSA
    Default Audio Device
dmix:CARD=ALSA,DEV=0
    bcm2835 ALSA, bcm2835 ALSA
    Direct sample mixing device
dmix:CARD=ALSA,DEV=1
    bcm2835 ALSA, bcm2835 IEC958/HDMI
    Direct sample mixing device
dsnoop:CARD=ALSA,DEV=0
    bcm2835 ALSA, bcm2835 ALSA
    Direct sample snooping device
dsnoop:CARD=ALSA,DEV=1
    bcm2835 ALSA, bcm2835 IEC958/HDMI
    Direct sample snooping device
hw:CARD=ALSA,DEV=0
    bcm2835 ALSA, bcm2835 ALSA
    Direct hardware device without any conversions
hw:CARD=ALSA,DEV=1
    bcm2835 ALSA, bcm2835 IEC958/HDMI
    Direct hardware device without any conversions
plughw:CARD=ALSA,DEV=0
    bcm2835 ALSA, bcm2835 ALSA
    Hardware device with all software conversions
plughw:CARD=ALSA,DEV=1
    bcm2835 ALSA, bcm2835 IEC958/HDMI
    Hardware device with all software conversions
default:CARD=Headset
    Logitech G35 Headset, USB Audio
    Default Audio Device
sysdefault:CARD=Headset
    Logitech G35 Headset, USB Audio
    Default Audio Device
front:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    Front speakers
surround21:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    IEC958 (S/PDIF) Digital Audio Output
dmix:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    Direct sample mixing device
dsnoop:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    Direct sample snooping device
hw:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    Direct hardware device without any conversions
plughw:CARD=Headset,DEV=0
    Logitech G35 Headset, USB Audio
    Hardware device with all software conversions
```

Parsing /proc/asound/cards isn't as straightforward as the others above (still no magic involved).
```
cat /proc/asound/cards 
 0 [ALSA           ]: bcm2835 - bcm2835 ALSA
                      bcm2835 ALSA
 1 [Headset        ]: USB-Audio - Logitech G35 Headset
                      Logitech Logitech G35 Headset at usb-3f980000.usb-1.5, full speed
```
For every device listed above there's a folder called `cardX`, e.g. card1 for the G35. Inside that folder are various other files and folders:
```
-r--r--r-- 1 root root 0 Dec  4 12:44 id
dr-xr-xr-x 3 root root 0 Dec  4 12:44 pcm0c
dr-xr-xr-x 3 root root 0 Dec  4 12:44 pcm0p
-r--r--r-- 1 root root 0 Dec  4 12:44 stream0
-r--r--r-- 1 root root 0 Dec  4 12:44 usbbus
-r--r--r-- 1 root root 0 Dec  4 12:44 usbid
-r--r--r-- 1 root root 0 Dec  4 12:44 usbmixer
```
The most interesting folders are the `pcmXp` folders (`pcmXc` are capture devices, in this case the headset's microphone) - inside there's a info file containing the following information:
```
card: 1
device: 0
subdevice: 0
stream: PLAYBACK
id: USB Audio
name: USB Audio
subname: subdevice #0
class: 0
subclass: 0
subdevices_count: 1
subdevices_avail: 1
```
The card and device part is important, as moonlight-embedded needs the audio device to be selected as `hw:X,Y` where X is the card number and Y is the device number, so in this case that's `hw:1,0`. As a side note, I'm not sure if something like `iec958:CARD=Headset,DEV=0` is supposed to work as parameter for moonlight-embedded as I don't know how it's handled internally and I've never tried either.

This information can obviously be obtained by looking at the output of `aplay -l` or `aplay -L` (and both would be faster and easier than walking dirs). The gist is: `aplay` provides a lot more information than parsing files listed above and I wasn't able to find this information elsewhere (yet).

If Kodi is listing more devices than Luna I'd suggest you have a look at your `/proc/asound/cards` - if the device is listed there as well, there's a huge chance that my RegEx is "swallowing" your device. In that case please paste the output of the above file so I can change the RegEx accordingly (did so a while back for HiFi Berries, just merged it into master 10 days back).

**Comment by wackerl91** [comment]
@tuzz 
Is there still something to be done / discussed here or can we close this issue? :)
